File: content-solutions/graph-advanced/mst-min-cable.java
import java.util.*;

class Solution {
    int[] parent, size;

    void init(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (x != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) {
            int t = ra;
            ra = rb;
            rb = t;
        }
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }

    public int solution(int n, int[][] cables) {
        init(n);
        int[][] sorted = cables.clone();
        Arrays.sort(sorted, (x, y) -> Integer.compare(x[2], y[2]));
        int total = 0, picked = 0;
        for (int[] e : sorted) {
            if (union(e[0], e[1])) {
                total += e[2];
                picked++;
            }
        }
        return picked == n - 1 ? total : -1;
    }
}

File: content-solutions/graph-advanced/mst-min-cable.py
def solution(n, cables):
    parent = list(range(n))
    size = [1] * n

    def find(x):
        root = x
        while parent[root] != root:
            root = parent[root]
        while x != root:
            nxt = parent[x]
            parent[x] = root
            x = nxt
        return root

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if size[ra] < size[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        size[ra] += size[rb]
        return True

    total = 0
    picked = 0
    for a, b, c in sorted(cables, key=lambda x: x[2]):
        if union(a, b):
            total += c
            picked += 1
    return total if picked == n - 1 else -1
